tasks/seal/task_submit_commit.go

```diff
@@ -158,116 +145,248 @@ func (s *SubmitCommitTask) Do(taskID harmonytask.TaskID, stillOwned func() bool)
 			pams = append(pams, *pam)
 		}
-	}
 
-	params.SectorActivations = append(params.SectorActivations, miner.SectorActivationManifest{
-		SectorNumber: abi.SectorNumber(sectorParams.SectorNumber),
-		Pieces:       pams,
-	})
-	params.SectorProofs = append(params.SectorProofs, sectorParams.Proof)
+		if sectorFailed {
+			break
+		}
 
-	enc := new(bytes.Buffer)
-	if err := params.MarshalCBOR(enc); err != nil {
-		return false, xerrors.Errorf("could not serialize commit params: %w", err)
+		ssize, err := pci.Info.SealProof.SectorSize()
+		if err != nil {
+			return false, xerrors.Errorf("could not get sector size: %w", err)
+		}
+
+		collateralPerSector, err := s.api.StateMinerInitialPledgeForSector(ctx, pci.Info.Expiration-ts.Height(), ssize, uint64(verifiedSize), ts.Key())
+		if err != nil {
+			return false, xerrors.Errorf("getting initial pledge collateral: %w", err)
+		}
+
+		collateralPerSector = big.Sub(collateralPerSector, pci.PreCommitDeposit)
+		if collateralPerSector.LessThan(big.Zero()) {
+			collateralPerSector = big.Zero()
+		}
+
+		collateral = big.Add(collateral, collateralPerSector)
+
+		params.SectorActivations = append(params.SectorActivations, miner.SectorActivationManifest{
+			SectorNumber: abi.SectorNumber(sectorParams.SectorNumber),
+			Pieces:       pams,
+		})
+		params.SectorProofs = append(params.SectorProofs, sectorParams.Proof)
+
+		sealedCID, err := cid.Parse(sectorParams.SealedCID)
+		if err != nil {
+			return false, xerrors.Errorf("parsing sealed CID: %w", err)
+		}
+
+		unsealedCID, err := cid.Parse(sectorParams.UnsealedCID)
+		if err != nil {
+			return false, xerrors.Errorf("parsing unsealed CID: %w", err)
+		}
+
+		infos = append(infos, proof.AggregateSealVerifyInfo{
+			Number:                abi.SectorNumber(sectorParams.SectorNumber),
+			Randomness:            sectorParams.TicketValue,
+			InteractiveRandomness: sectorParams.SeedValue,
+			SealedCID:             sealedCID,
+			UnsealedCID:           unsealedCID,
+		})
+
+		sectors = append(sectors, sectorParams.SectorNumber)
 	}
 
-	ssize, err := pci.Info.SealProof.SectorSize()
-	if err != nil {
-		return false, xerrors.Errorf("could not get sector size: %w", err)
+	if len(infos) > 1 && len(infos) < 4 {
+		// We have dropped some sectors and now don't have minimum, fail this task and send sectors back to pool
+		_, err = s.db.Exec(ctx, `UPDATE sectors_sdr_pipeline SET task_id_commit_msg = NULL WHERE task_id_commit_msg = $1 AND sp_id = $2 AND sector_number = ANY($3)`, taskID, sectorParamsArr[0].SpID, sectors)
+		if err != nil {
+			return false, xerrors.Errorf("unsetting commit task: %w", err)
+		}
+		return false, xerrors.Errorf("not enough sectors for aggregation: require 4 and have %d", len(infos))
 	}
 
-	collateral, err := s.api.StateMinerInitialPledgeForSector(ctx, pci.Info.Expiration-ts.Height(), ssize, uint64(verifiedSize), ts.Key())
-	if err != nil {
-		return false, xerrors.Errorf("getting initial pledge collateral: %w", err)
+	if len(infos) >= 4 {
```

Comment:
So before going into aggregate mode, we want to estimate gas to calculate the non-aggregated message fee. Then here we also would gas-estimate the aggregate version.

After we have the two estimates, we'd compare message costs (+aggregate fee batch balancer burn) and send the cheaper one